larq/optimizers.py

```diff
@@ -210,6 +210,16 @@ def _compute_var_opt_mapping(self, grads_and_vars):
                         f"No `default_optimizer` provided to train variable `{var}`."
                     )
 
+        print("!~!!!!")
+        print(self.var_opt_mapping)
+
+        # Make sure that each optimizer touches at least one variable
+        for optimizer_index, (_, optimizer) in enumerate(self.pred_opt_pairs):
+            if optimizer_index not in self.var_opt_mapping.values():
+                raise ValueError(
+                    f"Optimizer `{optimizer}` did not claim any variables."
+                )
+
```

Comment:
I think checking this in [`_apply_gradients()`](https://github.com/larq/larq/blob/c44ed976812edc2cc7d60c98155513942fc3af98/larq/optimizers.py#L136) would be simpler, then you could just do:
```python
for optimizer, opt_grads_and_vars in zip(self.optimizers, grad_var_lists):
    if not grad_var_lists:
        raise ValueError(f"Optimizer `{optimizer}` did not claim any variables.")
```

Reply:
Yes but then it happens on every training step; this way it just happens once when the var -> optimizer gets computed.

Reply:
TensorFlow also raises errors in `_apply_gradients` too so it wouldn't be problematic, but this is fine too.

Reply:
I meant it mostly for efficiency.